### PR TITLE
Upgrade host resolution

### PR DESCRIPTION
**Pull request by MatteoPiovanelli**
implements #8386 
I am proposing this for 1.10.x rather than dev because this introduces no breaking change in the way things are done: as it is, everything that is working right now, will keep working the same, but the new setting, when set to true, allows reversing some logic in selecting the tenant/shell that should respond to a request.

### CODE REVIEW

**Comment by MatteoPiovanelli** [comment]
@sebastienros  I made the changes I commented about on the issue. I am not merging this in yet because they are not the same we discussed yesterday. The result is the same, but I had to change a the test because it wasn't correctly testing for the conditions we have.

**Comment by MatteoPiovanelli** [comment]
Is this confirmed for me to merge in 1.10.x?

File: src/Orchard.Tests/Orchard.Framework.Tests.csproj [hunk 1]
```diff
@@ -459,4 +462,10 @@
   <Target Name="AfterBuild">
     <CallTarget Targets="CopySqlCeBinaries" />
   </Target>
+  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
```
Comment (sebastienros) [inline src/Orchard.Tests/Orchard.Framework.Tests.csproj]: Why is this change necessary?

**Review by sebastienros — APPROVED** [review]

File: src/Orchard.Tests/Orchard.Framework.Tests.csproj (hunk 1 above)
Comment (MatteoPiovanelli) [reply src/Orchard.Tests/Orchard.Framework.Tests.csproj]: it's not. VS had added those lines when I was updating the test adapters for the tests to work properly with VS 2017. I removed it in the latest push
